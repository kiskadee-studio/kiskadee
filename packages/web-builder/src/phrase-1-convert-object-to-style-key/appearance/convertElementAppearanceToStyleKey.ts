import type { Appearance, ClassNameMap, InteractionState, SingleColor } from '@kiskadee/schema';
import { elementClassNameMap, styleKeyUsageMap } from '../../utils';
import { generateCssRuleFromStyleKey } from '../../phrase-2-convert-style-key-to-css/generateCss';

function getShadowValue<T>(
  property: Partial<Record<InteractionState, T>>,
  state: InteractionState,
  defaultValue: T
): T {
  return property[state] !== undefined
    ? property[state]
    : property.rest !== undefined
      ? property.rest
      : defaultValue;
}

/**
 * Processes the given "appearance" object by extracting style-related properties
 * and generating keys for the `styleUsageMap`.
 *
 * Non-shadow properties are processed using the pattern:
 *   {property}__{value}
 * where if the value is an array, JSON.stringify is used.
 *
 * Shadow properties are processed as follows:
 *   - For the default state (rest): shadow__[x,y,blur,[color]]
 *   - For other states:             shadow--{state}__[x,y,blur,[color]]
 *
 * The numbers x, y, blur are included directly (without units) and the color,
 * which is of type SingleColor, is represented as a literal array.
 *
 * Note: Appearance never contains a "ref" key so that part is omitted.
 */
export function convertElementAppearanceToStyleKey(
  componentName: string,
  elementName: string,
  appearance: Appearance
) {
  let classNameMap = elementClassNameMap;

  // Process non-shadow properties
  for (const [key, value] of Object.entries(appearance)) {
    if (key.startsWith('shadow') === false) {
      if (
        typeof value === 'boolean' ||
        typeof value === 'string' ||
        typeof value === 'number' ||
        Array.isArray(value)
      ) {
        // Generate style key
        const formattedValue = Array.isArray(value) ? JSON.stringify(value) : value;
        const styleKey = `${key}__${formattedValue}`;
        styleKeyUsageMap[styleKey] = (styleKeyUsageMap[styleKey] || 0) + 1;

        // Map class name
        const { className } = generateCssRuleFromStyleKey(styleKey);
        const componentMap = classNameMap[componentName] ?? {};
        const elementMap = componentMap[elementName] ?? { rest: [] };
        elementMap?.rest?.push(className);
        const rest = elementMap.rest;

        classNameMap = {
          ...classNameMap,
          [componentName]: {
            ...componentMap,
            [elementName]: {
              ...elementMap,
              rest
            }
          }
        };
      }
    }
  }

  console.log({ nextState: JSON.stringify(classNameMap) });

  // Process shadow properties, if any exist.
  const hasShadowProperty =
    'shadowColor' in appearance ||
    'shadowX' in appearance ||
    'shadowY' in appearance ||
    'shadowBlur' in appearance;

  if (hasShadowProperty === true) {
    const { shadowX = {}, shadowY = {}, shadowBlur = {}, shadowColor = {} } = appearance;
    const allStates = new Set<InteractionState>([
      ...Object.keys(shadowX),
      ...Object.keys(shadowY),
      ...Object.keys(shadowBlur),
      ...Object.keys(shadowColor)
    ] as InteractionState[]);
    allStates.add('rest');

    for (const state of allStates) {
      // Generate style key
      const x = getShadowValue(shadowX, state, 0);
      const y = getShadowValue(shadowY, state, 0);
      const blur = getShadowValue(shadowBlur, state, 0);
      // Keep the color as a simple array (SingleColor)
      const color: SingleColor = getShadowValue(shadowColor, state, [0, 0, 0, 1]);
      // Create the key as a literal array string.
      const shadowKey =
        state === 'rest'
          ? `shadow__[${x},${y},${blur},${JSON.stringify(color)}]`
          : `shadow--${state}__[${x},${y},${blur},${JSON.stringify(color)}]`;
      styleKeyUsageMap[shadowKey] = (styleKeyUsageMap[shadowKey] || 0) + 1;

      // Map class name
      const { className } = generateCssRuleFromStyleKey(shadowKey);
      const componentMap = classNameMap[componentName] ?? {};
      const elementMap = componentMap[elementName] ?? {};
      const current = elementMap[state] ?? [];
      const nextElementMap = {
        ...elementMap,
        [state]: [...current, className]
      };

      classNameMap = {
        ...classNameMap,
        [componentName]: {
          ...componentMap,
          [elementName]: nextElementMap
        }
      };

      console.log({ nextState: JSON.stringify(classNameMap) });
    }
  }
}
