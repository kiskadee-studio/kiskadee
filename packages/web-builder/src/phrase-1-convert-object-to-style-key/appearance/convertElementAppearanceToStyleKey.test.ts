import type { Appearance } from '@kiskadee/schema';
import { describe, expect, it } from 'vitest';
import { convertElementAppearanceToStyleKey } from './convertElementAppearanceToStyleKey';

describe('convertElementAppearanceToStyleKey', () => {
  const component = 'button';
  const element = 'e1';

  describe('textItalic', () => {
    it('should generate textItalic true style key', () => {
      const appearance: Appearance = { textItalic: true };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        button: {
          e1: {
            rest: ['textItalic__true']
          }
        }
      });
    });

    it('should generate textItalic false style key', () => {
      const appearance: Appearance = { textItalic: false };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        button: {
          e1: {
            rest: ['textItalic__false']
          }
        }
      });
    });
  });

  describe('textWeight', () => {
    it('should generate textWeight thin style key', () => {
      const appearance: Appearance = { textWeight: 'thin' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__thin']
          }
        }
      });
    });

    it('should generate textWeight extraLight style key', () => {
      const appearance: Appearance = { textWeight: 'extraLight' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__extraLight']
          }
        }
      });
    });

    it('should generate textWeight light style key', () => {
      const appearance: Appearance = { textWeight: 'light' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__light']
          }
        }
      });
    });

    it('should generate textWeight normal style key', () => {
      const appearance: Appearance = { textWeight: 'normal' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__normal']
          }
        }
      });
    });

    it('should generate textWeight medium style key', () => {
      const appearance: Appearance = { textWeight: 'medium' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__medium']
          }
        }
      });
    });

    it('should generate textWeight semiBold style key', () => {
      const appearance: Appearance = { textWeight: 'semiBold' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__semiBold']
          }
        }
      });
    });

    it('should generate textWeight bold style key', () => {
      const appearance: Appearance = { textWeight: 'bold' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__bold']
          }
        }
      });
    });

    it('should generate textWeight extraBold style key', () => {
      const appearance: Appearance = { textWeight: 'extraBold' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__extraBold']
          }
        }
      });
    });

    it('should generate textWeight black style key', () => {
      const appearance: Appearance = { textWeight: 'black' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textWeight__black']
          }
        }
      });
    });
  });

  describe('textDecoration', () => {
    it('should generate textDecoration none style key', () => {
      const appearance: Appearance = { textDecoration: 'none' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textDecoration__none']
          }
        }
      });
    });

    it('should generate textDecoration underline style key', () => {
      const appearance: Appearance = { textDecoration: 'underline' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textDecoration__underline']
          }
        }
      });
    });

    it('should generate textDecoration line-through style key', () => {
      const appearance: Appearance = { textDecoration: 'lineThrough' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textDecoration__lineThrough']
          }
        }
      });
    });
  });

  describe('textAlign', () => {
    it('should generate textAlign left style key', () => {
      const appearance: Appearance = { textAlign: 'left' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textAlign__left']
          }
        }
      });
    });

    it('should generate textAlign center style key', () => {
      const appearance: Appearance = { textAlign: 'center' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textAlign__center']
          }
        }
      });
    });

    it('should generate textAlign right style key', () => {
      const appearance: Appearance = { textAlign: 'right' };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);
      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['textAlign__right']
          }
        }
      });
    });
  });

  describe('shadow', () => {
    it('should generate shadow style keys inheriting missing properties for rest and hover', () => {
      // "rest" defines all shadow properties.
      // "hover" only defines shadowX.
      // Expected: For hover, the missing properties should come from "rest"
      const appearance: Appearance = {
        shadowX: { rest: 10, hover: 20 },
        shadowY: { rest: 15 },
        shadowBlur: { rest: 5 },
        shadowColor: { rest: [0, 0, 0, 0.5] }
      };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['shadow__[10,15,5,[0,0,0,0.5]]'],
            hover: ['shadow--hover__[20,15,5,[0,0,0,0.5]]']
          }
        }
      });
    });

    it('should generate style key with shadow default values for rest and hover states', () => {
      // Only shadowX is set for "hover".
      // Expected: For "hover" state, shadowY and shadowBlur default to 0, and shadowColor defaults to [0,0,0,1].
      // Also, the "rest" state is processed with default values.
      const appearance: Appearance = {
        shadowX: { hover: 25 }
      };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['shadow__[0,0,0,[0,0,0,1]]'],
            hover: ['shadow--hover__[25,0,0,[0,0,0,1]]']
          }
        }
      });
    });

    it('should generate shadow style key with multiple interaction states and proper inheritance', () => {
      // "rest" defines complete shadow values.
      // "focus" defines its own shadowX and shadowY.
      // "hover" defines shadowY and shadowColor.
      // Expected: missing values inherit from "rest"
      const appearance: Appearance = {
        shadowX: { rest: 5, focus: 12 },
        shadowY: { rest: 8, focus: 16, hover: 10 },
        shadowBlur: { rest: 3 },
        shadowColor: { rest: [10, 20, 30, 0.8], hover: [50, 60, 70, 0.9] }
      };
      const result = convertElementAppearanceToStyleKey(component, element, appearance);

      expect(result).toEqual({
        [component]: {
          [element]: {
            rest: ['shadow__[5,8,3,[10,20,30,0.8]]'],
            focus: ['shadow--focus__[12,16,3,[10,20,30,0.8]]'],
            hover: ['shadow--hover__[5,10,3,[50,60,70,0.9]]']
          }
        }
      });
    });
  });
});
