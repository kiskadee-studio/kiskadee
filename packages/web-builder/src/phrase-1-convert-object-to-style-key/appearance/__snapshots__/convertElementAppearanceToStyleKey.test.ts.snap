// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertElementAppearanceToStyleKey > textAlign style key generation > generates the textAlign__center style key for center text alignment 1`] = `
[
  "textAlign__center",
]
`;

exports[`convertElementAppearanceToStyleKey > textAlign style key generation > generates the textAlign__left style key for left text alignment 1`] = `
[
  "textAlign__left",
]
`;

exports[`convertElementAppearanceToStyleKey > textAlign style key generation > generates the textAlign__right style key for right text alignment 1`] = `
[
  "textAlign__right",
]
`;

exports[`convertElementAppearanceToStyleKey > textDecoration style key generation > generates the textDecoration__lineThrough style key for line-through text decoration 1`] = `
[
  "textDecoration__lineThrough",
]
`;

exports[`convertElementAppearanceToStyleKey > textDecoration style key generation > generates the textDecoration__none style key for none text decoration 1`] = `
[
  "textDecoration__none",
]
`;

exports[`convertElementAppearanceToStyleKey > textDecoration style key generation > generates the textDecoration__underline style key for underline text decoration 1`] = `
[
  "textDecoration__underline",
]
`;

exports[`convertElementAppearanceToStyleKey > textItalic > should generate textItalic false style key 1`] = `"textItalic__false"`;

exports[`convertElementAppearanceToStyleKey > textItalic > should generate textItalic true style key 1`] = `"textItalic__true"`;

exports[`convertElementAppearanceToStyleKey > textItalic style key generation > generates the textItalic__false style key when italic is disabled 1`] = `"textItalic__false"`;

exports[`convertElementAppearanceToStyleKey > textItalic style key generation > generates the textItalic__true style key when italic is enabled 1`] = `"textItalic__true"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight black style key 1`] = `"textWeight__black"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight bold style key 1`] = `"textWeight__bold"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight extraBold style key 1`] = `"textWeight__extraBold"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight extraLight style key 1`] = `"textWeight__extraLight"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight light style key 1`] = `"textWeight__light"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight medium style key 1`] = `"textWeight__medium"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight normal style key 1`] = `"textWeight__normal"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight semiBold style key 1`] = `"textWeight__semiBold"`;

exports[`convertElementAppearanceToStyleKey > textWeight > should generate textWeight thin style key 1`] = `"textWeight__thin"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__black style key for black font weight 1`] = `"textWeight__black"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__bold style key for bold font weight 1`] = `"textWeight__bold"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__extraBold style key for extraBold font weight 1`] = `"textWeight__extraBold"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__extraLight style key for extraLight font weight 1`] = `"textWeight__extraLight"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__light style key for light font weight 1`] = `"textWeight__light"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__medium style key for medium font weight 1`] = `"textWeight__medium"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__normal style key for normal font weight 1`] = `"textWeight__normal"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__semiBold style key for semiBold font weight 1`] = `"textWeight__semiBold"`;

exports[`convertElementAppearanceToStyleKey > textWeight style key generation > generates the textWeight__thin style key for thin font weight 1`] = `"textWeight__thin"`;
